helm --wait option crash tiller server when deploying same chart.

After upgrading helm/tiller to latest version 2.2.1 , tiller started to crash whenever an existing deployment with the same chart gets deployed using the --wait option.

$ helm version
Client: &version.Version{SemVer:"v2.2.1", GitCommit:"db531fd75fb2a1fb0841a98d9e55c58c21f70f4c", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.2.1", GitCommit:"db531fd75fb2a1fb0841a98d9e55c58c21f70f4c", GitTreeState:"clean"}

Steps to reproduce:

1) create a new chart and deploy  (this works)
helm install -n my-test nginx-my-0.1.1.tgz --wait --debug --timeout 60

2) upgrade / rollback same version
helm upgrade my-test nginx-my-0.1.1.tgz --wait --debug --timeout 30
output:
Created tunnel using local port: '65109'
SERVER: "localhost:65109"
Error: UPGRADE FAILED: transport is closing

### container crashed
$ helm version
Client: &version.Version{SemVer:"v2.2.1", GitCommit:"db531fd75fb2a1fb0841a98d9e55c58c21f70f4c", GitTreeState:"clean"}
Error: cannot connect to Tiller

After that the pod just crashed and it recovers, or some times I had to kill the pod manually to get it back.

Do you think this is something that will be fix soon, or how long it will take to get the fix out there, I would like to have an estimate to relay in this new option. 
Thank you!


-------------
Tiller pod logs:

2017/02/28 21:34:48 storage.go:133: Getting release history for 'my-test'
2017/02/28 21:34:48 release_server.go:533: rolling back my-test (current: v2, target: v1)
2017/02/28 21:34:48 storage.go:37: Getting release "my-test" (v1) from storage
2017/02/28 21:34:48 release_server.go:913: Executing pre-rollback hooks for my-test
2017/02/28 21:34:48 release_server.go:940: Hooks complete for pre-rollback my-test
2017/02/28 21:34:48 client.go:393: Looks like there are no changes for nginx-my-deployment
2017/02/28 21:34:48 client.go:560: beginning wait for resources with timeout of 5m0s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x54b973]

goroutine 105 [running]:
panic(0x156c840, 0xc42000a040)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/helm/pkg/kube.(*Client).waitForResources.func1(0x14ce080, 0xc4207e4c60, 0xc420a5d5a7)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/kube/client.go:588 +0xe63
k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait.WaitFor(0xc420a2eee0, 0xc42091bb00, 0xc42006db00, 0x15daa00, 0xc420a2eee0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:271 +0x70
k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait.pollInternal(0xc420a2eee0, 0xc42091bb00, 0xc420a2eee0, 0xc42091bb00)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:187 +0x41
k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait.Poll(0x77359400, 0x45d964b800, 0xc42091bb00, 0x0, 0x1)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:183 +0x4d
k8s.io/helm/pkg/kube.(*Client).waitForResources(0xc420435660, 0x45d964b800, 0xc4209ab408, 0x1, 0x1, 0x1, 0x0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/kube/client.go:633 +0x167
k8s.io/helm/pkg/kube.(*Client).Update(0xc420435660, 0xc4203703e0, 0x1a, 0x1fd6d40, 0xc420245030, 0x1fd6d40, 0xc4202450a0, 0xc420176000, 0x12c, 0x1, ...)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/kube/client.go:267 +0x4cb
k8s.io/helm/pkg/tiller.(*ReleaseServer).performKubeUpdate(0xc4202e6020, 0xc420244bd0, 0xc420244fc0, 0x0, 0x12c, 0x1, 0xc4203703e0, 0x1a)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/tiller/release_server.go:510 +0x22a
k8s.io/helm/pkg/tiller.(*ReleaseServer).performRollback(0xc4202e6020, 0xc420244bd0, 0xc420244fc0, 0xc42012e5a0, 0x0, 0x0, 0x6)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/tiller/release_server.go:480 +0x181
k8s.io/helm/pkg/tiller.(*ReleaseServer).RollbackRelease(0xc4202e6020, 0x7fa9960e7ce8, 0xc42012e540, 0xc42012e5a0, 0x1ff0740, 0x18, 0x18)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/tiller/release_server.go:451 +0x7d
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_RollbackRelease_Handler.func1(0x7fa9960e7ce8, 0xc42012e540, 0x15b09a0, 0xc42012e5a0, 0xc420266340, 0xc42043e400, 0xc420335a90, 0x4110f8)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:900 +0xa2
k8s.io/helm/pkg/tiller.newUnaryInterceptor.func1(0x7fa9960e7ce8, 0xc42012e540, 0x15b09a0, 0xc42012e5a0, 0xc4202e7420, 0xc4202e7440, 0x15, 0xc420583b38, 0x0, 0xc420583b28)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/tiller/server.go:53 +0x18c
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_RollbackRelease_Handler(0x16c8d20, 0xc4202e6020, 0x7fa9960e7ce8, 0xc42012e540, 0xc420529770, 0x17e0a98, 0x0, 0x0, 0x2, 0x0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:902 +0x166
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc420266300, 0x1ff0740, 0xc42037c120, 0xc4203f03c0, 0xc4203ae450, 0x2037650, 0xc42012e4e0, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:638 +0xc14
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420266300, 0x1ff0740, 0xc42037c120, 0xc4203f03c0, 0xc42012e4e0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:796 +0x6b0
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc42043e300, 0xc420266300, 0x1ff0740, 0xc42037c120, 0xc4203f03c0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:449 +0xab
created by k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:450 +0xa3

--- a/pkg/kube/client.go
+++ b/pkg/kube/client.go
@@ -391,6 +391,11 @@ func updateResource(c *Client, target *resource.Info, currentObj runtime.Object,
 	}
 	if patch == nil {
 		log.Printf("Looks like there are no changes for %s", target.Name)
+		// This needs to happen to make sure that tiller has the latest info from the API
+		// Otherwise there will be no labels and other functions that use labels will panic
+		if err := target.Get(); err != nil {
+			return fmt.Errorf("error trying to refresh resource information: %v", err)
+		}
 		return nil
 	}
 

--- a/pkg/kube/client_test.go
+++ b/pkg/kube/client_test.go
@@ -147,14 +147,14 @@ func TestUpdate(t *testing.T) {
 	listB.Items[0].Spec.Containers[0].Ports = []api.ContainerPort{{Name: "https", ContainerPort: 443}}
 	listC.Items[0].Spec.Containers[0].Ports = []api.ContainerPort{{Name: "https", ContainerPort: 443}}
 
-	actions := make(map[string]string)
+	var actions []string
 
 	f, tf, codec, ns := cmdtesting.NewAPIFactory()
 	tf.Client = &fake.RESTClient{
 		NegotiatedSerializer: ns,
 		Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
 			p, m := req.URL.Path, req.Method
-			actions[p] = m
+			actions = append(actions, p+":"+m)
 			t.Logf("got request %s %s", p, m)
 			switch {
 			case p == "/namespaces/default/pods/starfish" && m == "GET":
@@ -201,16 +201,21 @@ func TestUpdate(t *testing.T) {
 	// if err := c.Update("test", objBody(codec, &listC), objBody(codec, &listA), false, 2, true); err != nil {
 	// 	t.Fatal(err)
 	// }
-	expectedActions := map[string]string{
-		"/namespaces/default/pods/dolphin":  "GET",
-		"/namespaces/default/pods/otter":    "GET",
-		"/namespaces/default/pods/starfish": "PATCH",
-		"/namespaces/default/pods":          "POST",
+	expectedActions := []string{
+		"/namespaces/default/pods/starfish:GET",
+		"/namespaces/default/pods/starfish:PATCH",
+		"/namespaces/default/pods/otter:GET",
+		"/namespaces/default/pods/otter:GET",
+		"/namespaces/default/pods/dolphin:GET",
+		"/namespaces/default/pods:POST",
 	}
-
+	if len(expectedActions) != len(actions) {
+		t.Errorf("unexpected number of requests, expected %d, got %d", len(expectedActions), len(actions))
+		return
+	}
 	for k, v := range expectedActions {
-		if m, ok := actions[k]; !ok || m != v {
-			t.Errorf("expected a %s request to %s", k, v)
+		if actions[k] != v {
+			t.Errorf("expected %s request got %s", v, actions[k])
 		}
 	}
 

